Build tx index in parallel with validation

jimpo opened the pull request:
This refactors the tx index code to be in it's own class and get built concurrently with validation code. The main benefit is decoupling and moving the txindex into a separate DB. The primary motivation is to lay the groundwork for other indexers that might be desired (such as the [compact filters](https://github.com/bitcoin/bips/pull/636)). The basic idea is that the TxIndex spins up its own thread, which first syncs the txindex to the current block index, then once in sync the BlockConnected ValidationInterface hook writes new blocks.

### DB changes

At the suggestion of some other developers, the txindex has been split out into a separate database. A data migration runs at startup on any nodes with a legacy txindex. Currently the migration blocks node initialization until complete.

### Open questions

- Should the migration of txindex data from the old DB to the new DB block in init or should it happen in a background thread? The downside to backgrounding it is that `getrawtransaction` would return an error message saying the txindex is syncing while the migration is running.

### Impact

In a sample size n=1 test where I synced nodes from scratch, the average time [Index writing](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L1903) was 3.36ms in master and 1.72ms in this branch. The average time between `UpdateTip` log lines for sequential blocks between 400,000 and IBD end on mainnet was 0.297204s in master and 0.286134s in this branch. Most likely this is just variance in IBD times, but I can try with some more trials if people want.

Commit:
[index] Create new TxIndex class.

The TxIndex will be responsible for building the transaction index
concurrently with the main validation thread by implementing
ValidationInterface. This does not process blocks concurrently yet.

Commit:
[index] TxIndex initial sync thread.

TxIndex starts up a background thread to get in sync with the block
index before blocks are processed through the ValidationInterface.

Commit:
[index] Allow TxIndex sync thread to be interrupted.

Commit:
[index] TxIndex method to wait until caught up.

In order to preserve getrawtransaction RPC behavior, there needs to be
a way for a thread to ensure the transaction index is in sync with the
current state of the blockchain.

Commit:
[init] Initialize and start TxIndex in init code.

Commit:
[validation] Replace tx index code in validation code with TxIndex.

Commit:
[rpc] Public interfaces to GetTransaction block until synced.

Now that the transaction index is updated asynchronously, in order to
preserve the current behavior of public interfaces, the code blocks
until the transaction index is caught up with the current state of the
blockchain.

Commit:
[test] Simple unit test for TxIndex.

promag commented on src/rest.cpp:
```diff
@@ -350,6 +351,10 @@ static bool rest_tx(HTTPRequest* req, const std::string& strURIPart)
         return RESTERR(req, HTTP_BAD_REQUEST, "Invalid hash: " + hashStr);
 
+    if (g_txindex) {
+        g_txindex->BlockUntilSyncedToCurrentChain();
```
Instead of blocking, give 503?

jimpo replied:
It's not really unavailable, just that the background process needs to catch up, which should happen quickly.

TheBlueMatt replied:
Should we not check the return of BlockUntilSyncedToCurrentChain and maybe always return an error before its even gotten close to caught up?

TheBlueMatt replied:
Still needs to be addressed.

jimpo replied:
https://github.com/bitcoin/bitcoin/pull/11857#issuecomment-377574766

jamesob commented on src/init.cpp:
```diff
@@ -1426,9 +1433,9 @@ bool AppInitMain(boost::thread_group& threadGroup, CScheduler& scheduler)
                 if (fRequestShutdown) break;
 
-                // LoadBlockIndex will load fTxIndex from the db, or set it if
+                // LoadBlockIndex will load fHavePruned if we've ever removed a
```
Comment incomplete or outdated?

jimpo replied:
Thanks, fixed.

TheBlueMatt commented:
High-Level Concept ACK. As for your two notes:

> an alternative would be to split the txindex into a separate DB and do a data migration on upgrade.

I'd vote strongly for this. Keeping them separate is good.

> though it may be fine to just do the TxIndex write directly in the BlockConnected method.

Yes, I think you should just do this. I'd like to move the CValidationInterface semantics to no longer be a single thread run on the scheduler but instead multiple scheduler threads that process events and only guarantee order for individual clients instead of globally.

promag commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,242 @@
+
+void TxIndex::Start()
+{
+    RegisterValidationInterface(this);
```
Move after Init?

jimpo replied:
No, this has to be registered before `m_synced` is set to true in Init.

Commit:
[db] Create separate database for txindex.

The new TxIndexDB class will be used by a future commit in this
change set.

Commit:
[db] Migration for txindex data to new, separate database.

laanwj commented:
Concept ACK

> an alternative would be to split the txindex into a separate DB and do a data migration on upgrade.

I agree that would be preferable (though not necessarily in this PR, I don't think the scope here should be extended further).  The transaction index has a completely different access pattern from the block index. This came up in #10922 and other places.

TheBlueMatt commented:
I'm not actaully sure that its unrelated - currently txindex is required to be kept in-sync and be present in the block tree DB before the coins for a block are flushed to the utxodb. Moving it to the background without bending over backwards to keep things in-sync would result in "corrupt" (I assume just missing entries) tx index in some cases on downgrade. Its not a huge deal, but I think moving the txindex into a separate DB would be really nice to do before (or in the same PR) things are background-flushed.

jimpo commented:
@TheBlueMatt @laanwj I updated this PR to split the database and added a migration. The migration took ~103 min on an AWS m4.large with data on a gp2 EBS volume (non-local SSD). So that's kind of painful. I could move the migration to happen in the background thread, but the UX might be weird for people upgrading because the RPC endpoint would report that the txindex is catching up. Open to suggestions here.

TheBlueMatt commented:
Seems to be missing migration (forget to push?). Doing migration in the background after start seems fine, better to wait with most things working than it just hanging startup, just have to make sure it doesn't overwrite new entries or you'd end up pointing to a reorg'd-out block's copy (though I think thats probably not technically a problem, it seems strange).

As for threading, it would be really nice to have the flushing for new blocks happen directly in the scheduler/validation interface threading, with the upgrade either in the init thread at the end (it just hangs around for the entire program's execution waiting for shutdown to start, so as long as the migration is interruptible that'd be a fine place for it) or a new thread.

jimpo commented:
@TheBlueMatt Migration is here: https://github.com/bitcoin/bitcoin/pull/11857/commits/a252d466ca0c44a455051d01c43e50e15643c423#diff-81e4f16a1b5d5b7ca25351a63d07cb80R446. Called from `Start()`, though it could be moved to `ThreadSync` (the dedicated thread for the txindex initial sync).

I also changed the new block flushing to happen directly in the `BlockConnected` callback as you suggested.

TheBlueMatt commented on src/txdb.cpp:
```diff
@@ -424,3 +425,129 @@ bool CCoinsViewDB::Upgrade() {
+            CompactRange(prev_key_newdb, key);
+            prev_key_newdb = key;
+        }
+        if (batch_olddb.SizeEstimate() > batch_size) {
```
Separating these if statements loses coherency - if we crash we may have a bogus txindex and no way to detect it. Should likely instead do if (batch_newdb.SizeEstimate() + batch_olddb.SizeEstimate() > batch_size)........

jimpo replied:
Good catch.

ryanofsky reviewed (commented):
Looks great! I reviewed most of this but rushed through the last few commits and want to look more closely before adding ack.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,129 @@
+        return false;
+    }
+
+    if (best_block_hash.IsNull()) {
```
Thanks, I moved the check for null `chain_tip` up to fix that case. However, I still think the early return here is necessary.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,129 @@
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint256, CDiskTxPos>> vPos;
+    vPos.reserve(block.vtx.size());
+    for (auto tx : block.vtx) {
```
Done

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,129 @@
+        }
+    } else {
+        if (best_block_index->GetAncestor(pindex->nHeight - 1) != pindex->pprev) {
+            FatalError("%s: Block %s does not connect to an ancestor of known best chain (tip=%s)",
```
This isn't checking against chainActive, it is just asserting that BlockConnected gets called with blocks in order (even if chainActive is ahead). It's important to note that if the BlockConnected callbacks are running, then `m_synced` is true an `ThreadSync` has exited. I added more comments to these fields/methods in the header file to hopefully make that more clear.

jimpo commented on src/txdb.cpp:
```diff
@@ -29,6 +29,7 @@ static const char DB_HEAD_BLOCKS = 'H';
 static const char DB_FLAG = 'F';
 static const char DB_REINDEX_FLAG = 'R';
 static const char DB_LAST_BLOCK = 'l';
+static const char DB_TXINDEX_BEST_BLOCK = 'T';
```
Yeah, that wasn't even used. Left over from a previous version where the database wasn't split out.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,107 @@ bool CCoinsViewDB::Upgrade() {
+bool TxIndexDB::WriteTxns(const std::vector<std::pair<uint256, CDiskTxPos>>& v_pos)
+{
+    CDBBatch batch(*this);
+    for (auto tuple : v_pos) {
```
Done.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,107 @@ bool CCoinsViewDB::Upgrade() {
+
+bool TxIndexDB::MigrateData(CBlockTreeDB& block_tree_db)
+{
+    // The prior implementation of txindex was always in sync with block index
```
I elaborated on the comment. I think this is the best way to determine whether a migration is necessary. LMK if you think it needs further explanation.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,107 @@ bool CCoinsViewDB::Upgrade() {
+    LogPrintf("Upgrading txindex database...\n");
+    LogPrintf("[0%%]...");
+    int report_done = 0;
+    size_t batch_size = 1 << 24;
```
Done.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,107 @@ bool CCoinsViewDB::Upgrade() {
+            break;
+        }
+
+        if (!pcursor->GetKey(key)) {
```
I think you are right that this should never happen, so I opted to return an error instead of logging and completing the migration.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,107 @@ bool CCoinsViewDB::Upgrade() {
+    WriteBatch(batch_newdb);
+    CompactRange(begin_key, key);
+    block_tree_db.WriteBatch(batch_olddb);
+    block_tree_db.CompactRange(begin_key, key);
```
Hmm, the logic gets kind of tricky since there's also the early break if the cursor is valid but iterates past the DB_TXINDEX range. Also the CompactRange after the loop compacts over the entire range, not just from the previous batch write point. What do you mean that it skips the last compact range on the old DB? It shouldn't.

This code was mostly copied from CCoinsViewDB if that wasn't clear.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11857#discussion_r159900328

> How do you feel about this approach?

Better, I'm still not sure why my suggestion doesn't work out, but this takes care of my concern.

jimpo commented on src/txdb.h:
```diff
@@ -142,9 +142,15 @@ class TxIndexDB : public CDBWrapper
     /// Migrate txindex data from the block tree DB, where it may be for older nodes that have not
     /// been upgraded yet to the new database.
-    bool MigrateData(CBlockTreeDB& block_tree_db);
+    bool MigrateData(CBlockTreeDB& block_tree_db, const uint256& block_hash);
```
It does get used. The `MigrateData` implementation is modified to write the block hash in this commit before deleting the txindex flag from the old DB.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,129 @@
+        vPos.push_back(std::make_pair(tx->GetHash(), pos));
+        pos.nTxOffset += ::GetSerializeSize(*tx, SER_DISK, CLIENT_VERSION);
+    }
+    return m_db->WriteTxns(vPos) && m_db->WriteBestBlockHash(pindex->GetBlockHash());
```
I kind of prefer to keep the DB writes together. I'd hoped that the access pattern would make it clear:

```c++
if (WriteBlock(*block, pindex)) {
    m_best_block_index = pindex;
}
```

laanwj commented:
Needs rebase (probably a simple one, only init.cpp conflicted).

Sjors reviewed (commented):
Concept ACK. It's very nice to able to add and remove txindex without a full reindex, which takes weeks on my Bitseed node. Worthy of a Performance label?

It tested (with testnet3) adding `txindex=1` on this branch, as well as adding `txindex=1` on master and then migrating using this branch.

Possibly unrelated, but upon switching back to `master`, deleting `/indexes`, launching `src/qt/bitcoin-qt -txindex=1`, confirming that I wanted to do a reindex, it failed with "Error opening block database". I then run `bitcoind -txindex=1 -reindex` for a minute, closed it and launched QT again, at which point indexing proceeded as expected.

In terms of storage something seems off (testnet3):

* chainstate without txindex: 968.7 MB
* chainstate with "legacy" txindex: 966.8 MB
* chainstate on this branch, after migration: 986.4 MB
* indexes, after migration: 870 MB
* indexes, without migration: 873 MB

So it looks like txindex only adds 30 MB on master, but the separate indexes directory adds almost 800 MB. Or I'm doing something wrong. 

Can you expand the progress logger beyond migration? It's also useful for when `txindex=1` is added after IBD.

Would be nice to have functional tests for the various `getrawtransaction` error messages.

It's a bit confusing to call the directory `/indexes` when the block index file isn't there, but maybe that can be moved later (or never, if it's too risky).

Sjors commented:
I also noticed that `indexes` isn't emptied if you set `txindex=0`. That's probably a good thing, e.g. perhaps someone doesn't want to read / update them for a while, but should be documented.

I was able to reproduce the "Error opening block database" QT error (on master) after a fresh reindex (on this branch). No need to use bitcoind to make it go away, simply launch with `-reindex`. That shouldn't be necessary. I'll check if this also happens on master.

Sjors commented:
I was comparing the wrong directory sizes above. `chainstate` is for the UTXO set, `txindex` is in `blocks/index` (on master).

Also the error I was seeing was unrelated, and hopefully fixed in #12401

jimpo commented on src/rpc/rawtransaction.cpp:
```diff
@@ -177,10 +183,12 @@ UniValue getrawtransaction(const JSONRPCRequest& request)
+        } else if (!g_txindex) {
+            errmsg = "No such mempool transaction. Use -txindex to enable blockchain transaction queries";
+        } else if (!f_txindex_ready) {
+            errmsg = "No such mempool transaction. Blockchain transactions are still in the process of being indexed";
```
Yes, that's right, an RPC consumer should wait until the index is built. Using the `RPC_INVALID_ADDRESS_OR_KEY` error code for the `"No such mempool transaction. Use -txindex to enable blockchain transaction queries"` clause also seems odd to me.

Perhaps the `RPC_IN_WARMUP` makes sense here, or perhaps a new `RPC_INDEX_UNAVAILABLE` error code.

TheBlueMatt replied:
A distinct error code may make sense, though no opinion on what. Further, I'm not sure we should let GetTransaction call into the g_txindex if !f_txindex_ready (as it may return data from a stale block, but its also just generally not so useful). Maybe GetTransaction/FindTx can check if we're in sync yet.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,132 @@ bool CCoinsViewDB::Upgrade() {
+            break;
+        }
+
+        // Log progress every 10%.
```
The percentage shown in the UI is updated with every percent, it's just the log line that shows increments of 10, as it appends to the same line: `[0%]...[10%]...[20%...]`. Alternatively, it could log progress on a separate line every 30 seconds or something like: "Upgrading txindex database: n% complete\n". I kind of prefer that approach, but I copied how it was done for CCoinsViewDB upgrade.

jimpo replied:
I opted to log every 30s.

jimpo commented:
@Sjors Thanks for the review!

I added f1b8b8d to log the txindex build status periodically while it is catching up.

I agree functional tests would be nice, but I'm having trouble figuring out how to exercise the case where the txindex is catchup up in the test harness. Could maybe sync a node with a few blocks, stop it, delete the txindex database files, restart and then hit the RPC, but it is going to be racey no matter what. Let me know if you have suggestions.

I'm curious what other people think, but having the block index database elsewhere doesn't bother me too much, because it is special and required. I am imagining that `indexes/` will be for optional indexes that can be built and synced in the background. Also, it doesn't seem worth it to make a migration just to move around the block index directory path.

Sjors commented:
@jnewbery: any thoughts on how to tackle functional tests?

eklitzke commented on src/init.cpp:
```diff
@@ -1568,6 +1574,12 @@ bool AppInitMain()
+
+        if (gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
+            auto txindex_db = MakeUnique<TxIndexDB>(nTxIndexCache, false, fReset);
+            g_txindex.reset(new TxIndex(std::move(txindex_db)));
```
This is a lot of moves, since you already move TxIndex::TxIndex. Can you change to create the unique_ptr in the constructor call, e.g.:

```
g_txindex.reset(new TxIndex(MakeUnique<...>(....));
```

Or even better, use argument forwarding.

jimpo replied:
I'd rather have them on separate lines because I think otherwise there is too much happening on one line (and it would get pretty long).

Regarding argument forwarding, even if `TxIndex` takes an rvalue-ref, the `std::move` is necessary.

eklitzke commented on src/index/txindex.h:
```diff
@@ -0,0 +1,74 @@
+                        const std::vector<CTransactionRef>& txn_conflicted) override;
+
+public:
+    explicit TxIndex(std::unique_ptr<TxIndexDB> db);
```
Can you make this forwardable?

```
TxIndex(std::unique_ptr<TxIndexDB> &&db);
```

Or just create the unique_ptr here, see my other comment regarding g_txindex.reset()

jimpo replied:
I find it clearer to pass by value here, as is recommended by [this StackOverflow post](https://stackoverflow.com/a/8114913). Is there a big benefit to changing the parameter to an rvalue-ref?

eklitzke replied:
You're right, disregard.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,251 @@
+        }
+    }
+
+    SyncWithValidationInterfaceQueue();
```
I think it makes sense to add a logging statement in the path here where the method actually blocks.

jimpo replied:
Done.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,251 @@
+    }
+
+    m_thread_sync = std::thread(&TraceThread<std::function<void()>>, "txindex",
+                                std::function<void()>(std::bind(&TxIndex::ThreadSync, this)));
```
Doesn't std::bind already return the right type here?

jimpo replied:
Good catch. This came from a copy-paste.

eklitzke commented on src/txdb.cpp:
```diff
@@ -424,3 +424,132 @@ bool CCoinsViewDB::Upgrade() {
+
+bool TxIndexDB::ReadTxPos(const uint256 &txid, CDiskTxPos& pos) const
+{
+    return Read(std::make_pair(DB_TXINDEX, txid), pos);
```
I believe you can do `{DB_TXINDEX, txid}` in C++11.

jimpo replied:
That doesn't seem to work because `Read` is templated and can't infer that the initializer list should be cast to a `std::pair`.

eklitzke reviewed (commented):
Mostly style stuff, will do another pass later to review the logic more closely.

sipa commented:
Concept ACK.

Having the transaction index as a modular, optional, separate database that gets updated in the background is a major improvement over how things work now.

It's also the approach that similar optional indexes or redundant data can take; for example a rolling UTXO set hash or a per-scriptPubKey UTXO index (see #9806) could follow the same approach.

Sjors commented:
Tested adding an index in Bitcoin-QT and it seems to work. I didn't try upgrading an existing index.

It doesn't delete the index when you set `txindex=0`. That might actually be better than the current behavior to prevent accidents, but should be explained in the release notes and such.

jimpo commented:
@Sjors Thanks for testing. I fixed the issue switching from pruned to unpruned w/o txindex with an amend to 33ebb98. You raise an interesting point about dropping the txindex database. Not sure what the right behavior should be. One approach might be to add a `-dropindex=txindex` flag or to add an RPC and `bitcoin-cli` command that errors if the txindex is active. I like the CLI approach personally, but the downside is it requires the RPC server to be running.

Commit:
[doc] Include txindex changes in the release notes.

Sjors commented:
I think your release note is fine. Deleting a directory doesn't seem worth the complexity of another RPC command / command line flag. I don't think it's too much to ask.

jonasschnelli commented:
Concept ACK!
- Is there a reason for the extensive use of the c++11 `auto` specifier (for things like `CBlockIndex*`)?
- Maybe a stupid question, but why blocking the thread / RPC / REST during txindex generation? I just had the thought why not building the txindex reserve (from tip to genesis) and allow access anytime while eventually add a txindex status report call `txindexstatus` (if this makes sense, then probably not in this PR).

jimpo commented:
@jonasschnelli I wrote this before the guidance on when to use `auto` was merged into the style docs. I'll go back through the commits and replace `auto` with concrete types accordingly.

As for the RPC behavior, it actually works similar to how you suggest. `BlockUntilSyncedToCurrentChain` only blocks if the txindex is caught up and just waits for the ValidationInterface queue to clear. If the txindex is syncing from genesis, `BlockUntilSyncedToCurrentChain` immediately returns false. The `getrawtransaction` behavior, however, is to optimistically query the index and only if it can't find the tx does it return a not found error saying the index is catching up. Basically, the index has an initial sync phase, then a staying in sync phase, where RPC calls might block for a short amount of time just for consistency with other RPC responses.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+#include <validation.h>
+#include <warnings.h>
+
+static constexpr int64_t SYNC_LOG_INTERVAL = 30; // seconds
```
remove `static`, that's the opposite of what you want (it reserves `.bss` space)

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+
+    // Attempt to migrate txindex from the old database to the new one. Even if
+    // chain_tip is null, the node could be reindexing and we still want to
+    // delete txindex records the old database.
```
typo, "in the old database"

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+    if (pindex) {
+        LogPrintf("txindex is enabled at height %d\n", pindex->nHeight);
+    } else {
+        LogPrintf("txindex is enabled\n", pindex->nHeight);
```
This is wrong, you have an extra argument that dereferences a nullptr.

jimpo replied:
Thanks.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+{
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint256, CDiskTxPos>> vPos;
+    vPos.reserve(block.vtx.size());
```
nit: move this to the ctor (it can take a size to reserve)

jimpo replied:
I'm pretty sure the `std::vector` constructor that takes a `size_t` creates the vector with `n` elements rather than having a size of 0 and a capacity of `n`.

eklitzke replied:
That's not strictly accurate:

```
$ g++ --version
g++ (GCC) 7.3.1 20180303 (Red Hat 7.3.1-5)
Copyright (C) 2017 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

$ cat reserve.cc 
#include <iostream>
#include <vector>

struct Foo {
    int x;
    Foo() :x(1) {}
};

int main(int argc, char **argv) {
    std::vector<Foo> foovec;
    foovec.reserve(10);
    std::cout << foovec[5].x << std::endl;
    std::cout << foovec.at(5).x << std::endl;
    return 0;
}

$ g++ reserve.cc 

$ ./a.out 
0
terminate called after throwing an instance of 'std::out_of_range'
  what():  vector::_M_range_check: __n (which is 5) >= this->size() (which is 0)
Aborted (core dumped)
```

In this test `Foo` has a constructor that sets `x` to 1, but as you can see from the first line of output it prints 0 implying that the ctor wasn't actually called. The `.at()` check fails because the element is out of bounds compared to the true size. The standard calls this ["default insertable"](http://en.cppreference.com/w/cpp/concept/DefaultInsertable) which I think just means "not allocator aware" in which case the default allocator is allowed to just create space for the item.

Either way it's a nit that you can ignore, but I think the suggestion I made is safe and slightly faster, assuming you actually initialize the elements before accessing them.

jimpo replied:
Have you tried that code sample with `std::vector<Foo> foovec(10);` instead of `std::vector<Foo> foovec; foovec.reserve(10);`? The behavior is different. I assume that is the suggestion you are making, though maybe I misunderstood.

jimpo replied:
Have you tried that code sample with std::vector<Foo> foovec(10); instead of std::vector<Foo> foovec; foovec.reserve(10);? The behavior is different. I assume that is the suggestion you are making, though maybe I misunderstood.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+    std::vector<std::pair<uint256, CDiskTxPos>> vPos;
+    vPos.reserve(block.vtx.size());
+    for (const auto& tx : block.vtx) {
+        vPos.push_back(std::make_pair(tx->GetHash(), pos));
```
nit: emplace_back

eklitzke commented on src/txdb.h:
```diff
@@ -126,4 +126,31 @@ class CBlockTreeDB : public CDBWrapper
+    bool ReadTxPos(const uint256& txid, CDiskTxPos& pos) const;
+
+    /// Write a batch of transaction positions to the DB.
+    bool WriteTxns(const std::vector<std::pair<uint256, CDiskTxPos>>& v_pos);
```
IMO it's weird to mix tx and txn, I would just use tx (and txs) throughout. Up to you though, purely a style thing.

jimpo replied:
I agree, will change.

eklitzke commented on src/txdb.cpp:
```diff
@@ -424,3 +424,132 @@ bool CCoinsViewDB::Upgrade() {
+        batch_olddb.Erase(key);
+
+        if (batch_newdb.SizeEstimate() > batch_size || batch_olddb.SizeEstimate() > batch_size) {
+            WriteBatch(batch_newdb);
```
The transaction logic throughout is wrong here because there's a data ordering dependency.

During the data migration:
 * Write to new new with `fsync = True`
 * After new write finishes, you can erase from old db (fsync setting can be left false)

This ensures that you can can always scan through the old database and catch up to what's in the new database if there's a crash or you are interrupted. During the non-migration path I think it's fine to have fsync disabled.

eklitzke commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,252 @@
+        vPos.push_back(std::make_pair(tx->GetHash(), pos));
+        pos.nTxOffset += ::GetSerializeSize(*tx, SER_DISK, CLIENT_VERSION);
+    }
+    return m_db->WriteTxns(vPos) && m_db->WriteBestBlockHash(pindex->GetBlockHash());
```
These should both be done in a single `WriteBatch()` call to make them atomic.

jimpo replied:
Will do.

eklitzke reviewed (changes requested):
This looks pretty good. There are some data consistency issues and a null dereference I spotted, but they should be easy to fix.

eklitzke commented:
I thought more about the ordering dependency between the block index and the txindex, and I realize now that it might be possible to make them independent of ordering. There's an edge case (that may not be possible in practice) if a reorg happens during crash recovery, so I stand by my original recommendation and maybe we can fix the reorg edge case before 0.17 is released.

ryanofsky reviewed (commented):
utACK 21f8a3bae27ffbbfe9e5d8a6d3c6b4e8c09045e1. This is really nicely implemented. I left some suggestions below, but they are mostly requests to add comments.

* On the open question in the PR description: I don't understand the downside to migrating in a background thread. Isn't getrawtransaction unusable during migration whether it's happening in the  foreground or background?

* Other things could be cleaned up in the PR description: "its" spelled "it's," repetition of "The primary motivation is to lay the groundwork for other indexers" and "The real goal is to lay the groundwork for other indexers"

* @eklitzke, can you clarify what the broken "reorg edge case" is from
https://github.com/bitcoin/bitcoin/pull/11857#issuecomment-372559696, so we can at least document it, if not fix it here.

ryanofsky commented:
Two functional tests `rpc_rawtransaction.py` and `wallet_abandonconflict.py` seem to be failing, perhaps due to a bad interaction with #11041. AssertLockHeld is failing with stack trace:

```
#0  0x00007ffff5411428 in __GI_raise (sig=sig@entry=6) at ../sysdeps/unix/sysv/linux/raise.c:54
#1  0x00007ffff541302a in __GI_abort () at abort.c:89
#2  0x0000555555b617d0 in AssertLockHeldInternal (pszName=0x555555bfe1ff "cs_main", pszFile=0x555555bfe1f0 "./validation.h", nLine=432, cs=0x55555615d240 <cs_main>) at sync.cpp:157
#3  0x00005555555beae4 in LookupBlockIndex (hash=...) at ./validation.h:432
#4  0x00005555557a487f in TxToJSON (tx=..., hashBlock=..., entry=...) at rpc/rawtransaction.cpp:52
#5  0x00005555557a5835 in getrawtransaction (request=...) at rpc/rawtransaction.cpp:200
#6  0x00005555557d07d6 in CRPCTable::execute (this=0x555556166d80 <tableRPC>, request=...) at rpc/server.cpp:493
#7  0x000055555596ffae in HTTPReq_JSONRPC (req=0x7fffcc0017b0) at httprpc.cpp:188
#8  0x00005555557542ae in std::_Function_handler<bool (HTTPRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&), bool (*)(HTTPRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>::_M_invoke(std::_Any_data const&, HTTPRequest*&&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) (__functor=..., __args#0=<unknown type in /home/russ/src/bitcoin/src/bitcoind, CU 0x509db3, DIE 0x562965>, __args#1="")
    at /usr/include/c++/5/functional:1857
#9  0x000055555597d64a in std::function<bool (HTTPRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&)>::operator()(HTTPRequest*, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const (
    this=0x7fffcc001c10, __args#0=0x7fffcc0017b0, __args#1="") at /usr/include/c++/5/functional:2267
#10 0x000055555597b87b in HTTPWorkItem::operator() (this=0x7fffcc001be0) at httpserver.cpp:53
#11 0x000055555597e7c4 in WorkQueue<HTTPClosure>::Run (this=0x5555562e3ea0) at httpserver.cpp:112
#12 0x00005555559768d7 in HTTPWorkQueueRun (queue=0x5555562e3ea0) at httpserver.cpp:333
#13 0x000055555598ed42 in std::_Bind_simple<void (*(WorkQueue<HTTPClosure>*))(WorkQueue<HTTPClosure>*)>::_M_invoke<0ul>(std::_Index_tuple<0ul>) (this=0x5555562e4368) at /usr/include/c++/5/functional:1531
#14 0x000055555598e6f8 in std::_Bind_simple<void (*(WorkQueue<HTTPClosure>*))(WorkQueue<HTTPClosure>*)>::operator()() (this=0x5555562e4368) at /usr/include/c++/5/functional:1520
#15 0x000055555598de99 in std::thread::_Impl<std::_Bind_simple<void (*(WorkQueue<HTTPClosure>*))(WorkQueue<HTTPClosure>*)> >::_M_run() (this=0x5555562e4350) at /usr/include/c++/5/thread:115
#16 0x00007ffff5d82510 in ?? () from /usr/lib/x86_64-linux-gnu/libstdc++.so.6
#17 0x00007ffff6d556ba in start_thread (arg=0x7fffc7fff700) at pthread_create.c:333
#18 0x00007ffff54e341d in clone () at ../sysdeps/unix/sysv/linux/x86_64/clone.S:109
```

The error is showing up on travis (https://travis-ci.org/bitcoin/bitcoin/jobs/353440060), but I can also reproduce the problem locally building with building with `--enable-debug`.

jimpo commented on src/index/txindex.h:
```diff
@@ -0,0 +1,52 @@
+
+/**
+ * TxIndex is used to look up transactions included in the blockchain by hash.
+ * The index is written to a keyspace in the block index database and records
```
Yeah, outdated comment, will fix. An earlier version of this refactor didn't split the DB out, just the logic.

jimpo commented on src/txdb.h:
```diff
@@ -126,4 +126,25 @@ class CBlockTreeDB : public CDBWrapper
 };
 
+/** Access to the block database (blocks/index/) */
+class TxIndexDB : public CDBWrapper
```
Done.

jimpo commented on src/txdb.h:
```diff
@@ -126,4 +126,25 @@ class CBlockTreeDB : public CDBWrapper
+public:
+    explicit TxIndexDB(size_t n_cache_size, bool f_memory = false, bool f_wipe = false);
+
+    TxIndexDB(const TxIndexDB&) = delete;
```
Done.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,113 @@ bool CCoinsViewDB::Upgrade() {
+    std::pair<unsigned char, uint256> prev_key_olddb = begin_key;
+
+    std::unique_ptr<CDBIterator> pcursor(block_tree_db.NewIterator());
+    for (pcursor->Seek(begin_key); pcursor->Valid(); pcursor->Next()) {
```
Done.

jimpo commented on src/index/txindex.h:
```diff
@@ -0,0 +1,52 @@
+                        const std::vector<CTransactionRef>& txn_conflicted) override;
+
+public:
+    explicit TxIndex(std::unique_ptr<TxIndexDB> db);
```
Done.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,138 @@
+bool TxIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
+{
+    CDiskTxPos pos(pindex->GetBlockPos(), GetSizeOfCompactSize(block.vtx.size()));
+    std::vector<std::pair<uint256, CDiskTxPos>> vPos;
```
Yes, it could be more efficient, but I feel that this produces the cleanest separation of responsibilities. Another way to do it might be to add a `TxIndexDB::WriteTxToBatch(CDBBatch& batch, const uint256& txid, const CDiskTxPos& tx_pos)` method. Would you prefer that?

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11857#discussion_r174537635

> Yes, it could be more efficient, but I feel that this produces the cleanest separation of responsibilities.

I'm not sure what the separation of responsibilities is actually. If it's important to avoid a more direct write, can you say somewhere in a comment what the separation is supposed to be, and what benefits it provides? This just seems a little more complicated than I would expect it to be.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,138 @@
+        return;
+    }
+
+    // Ensure block connects to an ancestor of the current best block.
```
I'll elaborate on the comment, but I don't understand the case you are describing. These checks should always pass as long as the ValidationInterface BlockConnected notifications are sent in order. If there is a race you can think of in the presence of reorgs or shutdowns, we should fix that.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11857#discussion_r174577538

> I'll elaborate on the comment, but I don't understand the case you are describing

The `BlockConnected` handler will drop any notifications as long as `m_synced` is false, but when `m_synced` becomes true, there is no code to clear out the queue, which could be backlogged. So if `m_synced` is set to true after a reorg but before `BlockConnected` calls from the reorg arrive, some initial `BlockConnected` notifications about blocks that aren't descendants of the synced best block could fail.

jimpo replied:
Ah, got it. Will update the comment and make this a warning log instead of a fatal error.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -10,6 +11,8 @@
 #include <validation.h>
 #include <warnings.h>
 
+constexpr int64_t SYNC_LOG_INTERVAL = 30; // seconds
```
Correct me if I'm wrong, but I assume this could be inlined or something by the compiler if it were not declared static. I tried looking it up briefly and it said something along those lines, but then I got bored because I don't care all that much (something, something ODR-rule?). Will change if you still want.

ryanofsky replied:
https://github.com/bitcoin/bitcoin/pull/11857#discussion_r174594791

> Correct me if I'm wrong, but I assume this could be inlined or something by the compiler if it were not declared static. I tried looking it up briefly and it said something along those lines, but then I got bored because I don't care all that much (something, something ODR-rule?). Will change if you still want.

Hmm, I guess you @eklitzke are making me question my sanity, but why would adding `static` here prevent inlining, or "reserve .bss space" or do anything along these lines?

If this isn't declared static, the compiler doesn't know that there isn't some code in another object file that will access the SYNC_LOG_INTERVAL variable or take its address. So there will necessarily be a SYNC_LOG_INTERVAL symbol in the object file. If it's static, it can just be internal to the object.

jimpo commented on src/index/txindex.h:
```diff
@@ -56,6 +56,10 @@ class TxIndex final : public CValidationInterface
 
+    /// Blocks the current thread until the transaction index is caught up to
+    /// the current state of the block chain.
+    bool BlockUntilSyncedToCurrentChain();
```
Good point, updated comment.

jimpo commented on src/init.cpp:
```diff
@@ -1598,15 +1610,22 @@ bool AppInitMain()
+    // ********************************************************* Step 8: start indexers
+    if (gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX)) {
+        auto txindex_db = MakeUnique<TxIndexDB>(nTxIndexCache, false, fReindex);
+        g_txindex.reset(new TxIndex(std::move(txindex_db)));
```
Done.

jimpo commented on src/init.cpp:
```diff
@@ -1404,16 +1411,22 @@ bool AppInitMain()
     nTotalCache = std::min(nTotalCache, nMaxDbCache << 20); // total cache cannot be greater than nMaxDbcache
-    int64_t nBlockTreeDBCache = nTotalCache / 8;
-    nBlockTreeDBCache = std::min(nBlockTreeDBCache, (gArgs.GetBoolArg("-txindex", DEFAULT_TXINDEX) ? nMaxBlockDBAndTxIndexCache : nMaxBlockDBCache) << 20);
+    int64_t nBlockTreeDBCache = nTotalCache / 16;
```
You're right, will revert. I was just kind of making up values here.

ryanofsky commented:
I'll take another look at this tomorrow to re-ack the new changes. Hopefully I responded to wherever there were questions above.

jimpo commented on src/test/txindex_tests.cpp:
```diff
@@ -0,0 +1,47 @@
+
+    for (const auto& txn : coinbaseTxns) {
+        CDiskTxPos postx;
+        BOOST_CHECK(txindex.FindTx(txn.GetHash(), postx));
```
Done.

jimpo commented on src/test/txindex_tests.cpp:
```diff
@@ -0,0 +1,47 @@
+        txindex.BlockUntilSyncedToCurrentChain();
+
+        CDiskTxPos actual_postx;
+        BOOST_CHECK(txindex.FindTx(txn.GetHash(), actual_postx));
```
I added checks for 1 & 3. With regards to persistence, the DB in the test environment currently uses the in-memory LevelDB environment rather than actually writing to disk, which makes cleanup nicer and probably makes the test marginally faster.

How important do you think it is to test the persistence at this layer? My feeling is that it's probably not adding much since it's just using the common CDBWrapper abstraction.

jnewbery commented:
@Sjors
> @jnewbery: any thoughts on how to tackle functional tests?

My first thought is that this isn't suitable for a functional test, since it's an implementation change rather than new functionality. Unit tests seem more appropriate.

Will try to review this week.

sipa commented:
This seems to introduce a cyclic dependency between `index/txindex` and `validation`. As these usually are a sign of imperfect abstraction boundaries, I would very much like to avoid them. If the current code is only expected to be temporary, I don't want to hold things up just for this, but I don't think it's very hard to avoid; you can move `GetTransaction` from `validation` to `rpc/rawtransaction` or another utility file. It doesn't quite belong in `validation` anyway.

jimpo commented:
@sipa Totally agree `GetTransaction` should be moved. I'd prefer to do it in a follow-up PR as this one is already rather large though.

TheBlueMatt commented on src/index/txindex.cpp:
```diff
@@ -13,6 +13,8 @@
 
 constexpr int64_t SYNC_LOG_INTERVAL = 30; // seconds
 
+std::unique_ptr<TxIndex> g_txindex;
```
It'd be nice to keep these kinds of globals in the same place. At least g_connman is in init.cpp and exposed as extern in net.h, so you could duplicate that (which is "correct" in that init "owns" the object).

jimpo replied:
That's what I did originally, but that caused problems when building tests. The test binary is not linked against init.cpp, it's linked against test_bitcoin_main.cpp, which is why `g_connman` is also defined there. That took me an annoying amount of time to figure out, and by the time I did I was so pissed off I decided to move it to `txindex.cpp`.

This whole thing where stuff in init has to get redefined in `test_bitcoin_main.cpp` if it's used in the tests seems pretty gross. Do you really thing that's the right way to do it?

TheBlueMatt replied:
Hmm, yea, thats a bit gross. Still, if thats what we have now, maybe add a comment noting that in both places and at least be consistent?

TheBlueMatt commented on src/rpc/rawtransaction.cpp:
```diff
@@ -229,29 +238,36 @@ UniValue gettxoutproof(const JSONRPCRequest& request)
 
+    // Allow txindex to catch up if we need to query it and before we acquire cs_main.
+    if (g_txindex && !pblockindex) {
+        g_txindex->BlockUntilSyncedToCurrentChain();
```
Also here - should report a different error of some kind if we're not in sync yet.

jimpo replied:
The error here already is not that useful and does not even indicate if the txindex is disabled. I'd like to fix this in a follow-up PR.

TheBlueMatt reviewed (commented):
Didnt review tests, also note that doc/files.md should be updated with new datadir paths.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,139 @@
+
+    const CBlockIndex* pindex = LookupBlockIndex(best_block_hash);
+    if (!pindex) {
+        FatalError("%s: Last block synced by txindex is unknown", __func__);
```
I don't understand. This line is in the `Init` method, not the BlockConnected callback. As long as we never prune entries from the block index, I don't see when this check would fail.

TheBlueMatt replied:
A block may be connected, but the block index may not be written to disk until much later. Thus, we can run far ahead here.

jimpo replied:
See last two commits (c837c6c & 297f89c) for fix.

sipa commented:
I'd very much like to get rid of the UTXO based block lookup. It's unreliable (depends on having at least one unspent output left for the transaction), is slow (needs to load and scan the entire block) and probably unused. There is an issue for it somewhere, but I think we'll want a deprecation for it first.

One way to deal with it is introducing a new RPC `getchaintransaction` or so, which strictly only works when txindex is enabled. Then `getrawtransaction` can be deprecated with a warning saying you can use `getmempoolentry` for unconfirmed transactions, and `getchaintransaction` for confirmed ones (if txindex is enabled).

Commit:
[index] Move disk IO logic from GetTransaction to TxIndex::FindTx.

jimpo commented on src/txdb.cpp:
```diff
@@ -458,3 +458,116 @@ bool TxIndexDB::ReadBestBlockHash(uint256& hash) const {
+    const size_t batch_size = 1 << 24; // 16 MiB
+
+    CDBBatch batch_newdb(*this);
+    CDBBatch batch_olddb(block_tree_db);
```
What do you think of this approach? https://github.com/bitcoin/bitcoin/pull/11857/commits/c3b34b1b62ab034c5d6b6d700ae378291b19fc4b

jimpo commented:
@TheBlueMatt @sipa I'd prefer to handle some of your suggestions about getrawtransaction (better error reporting, no UTXO lookup, moving GetTransaction out of validation, etc.) in a follow-up PR because this one is already 700+ lines. Commit https://github.com/jimpo/bitcoin/commit/2b8ee538bf804bcd2d30942d1254d147e65f2ee3 is a starting point.

sipa commented:
@jimpo Sorry if I was unclear; my message above was just to give some background thoughts and not a request to further change this PR itself.

I'll review the code soon.

ryanofsky commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,271 @@
+    }
+
+    // Ensure block connects to an ancestor of the current best block. This should always be the
+    // case assuming BlockConnected is called on the ValidationInterfaces in the proper order.
```
In commit "[index] Create new TxIndex class."

It still seems to me like it would be possible (though unlikely) for these errors to trigger spuriously, even when the events arrive in the right order, but the notification queue gets backlogged: https://github.com/bitcoin/bitcoin/pull/11857#discussion_r175614840

If this is true, the current comment seems misleading, and could mention the edge case.

jimpo replied:
Fixed the comment and error handling.

ryanofsky reviewed (commented):
utACK 297f89cad4808b3499d290bdadb5c4863a94fdd7 with squashed commits. Lots of updates since my previous review, too many to mention, but all in response to comments above.

Sjors commented:
Tested 806b2f1, including interrupting the migration from legacy to the new db.

If someone wants to do a partial backport, the migration code that removes txindex without requiring a reindex would be useful for folks who regret having set `txindex=1` on a node with slow hardware.

ryanofsky reviewed (commented):
utACK 806b2f1764b6e5a9c7abec887bfa89cd386648d8. Only change since last review is rebase and removing "Last block synced by txindex is unknown" error.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -195,6 +183,16 @@ void TxIndex::BlockConnected(const std::shared_ptr<const CBlock>& block, const C
+    if (!m_synced) {
+        return;
+    }
+    if (!m_db->WriteBestBlock(locator)) {
```
Yeah, that's a good idea.

jtimon commented:
Concept ACK

ryanofsky commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,294 @@
+
+void TxIndex::Stop()
+{
+    UnregisterValidationInterface(this);
```
@jamesob got me looking into this code, but since Stop() is called from the destructor, is it ok to call this here, or is there a problem like https://github.com/bitcoin/bitcoin/pull/12647#discussion_r179790601?

jamesob reviewed (commented):
Partially through a review: I've finished looking at `TxIndex` and some related parts of `TxIndexDB`. Looks great so far, very excited for this change.

jimpo commented on src/txdb.h:
```diff
@@ -126,4 +124,36 @@ class CBlockTreeDB : public CDBWrapper
+    bool ReadBestBlock(CBlockLocator& hash) const;
+
+    /// Write block locator of the chain that the txindex is in sync with.
+    bool WriteBestBlock(const CBlockLocator& locator);
```
So `TxIndex::Init()` uses `FindForkInGlobalIndex` to populate `m_best_block_index`, which may not necessarily be the tip of the locator. One case where it might not be is when the active chain locator is written at the end of ThreadSync, which may contain locator entries for blocks that have not been flushed to disk.

jimpo commented on src/index/txindex.cpp:
```diff
@@ -0,0 +1,294 @@
+        return;
+    }
+
+    if (!m_db->WriteBestBlock(locator)) {
```
Yeah, there's a few reasons. 1) It's how CWallet works, which I looked at as a model. 2) I'd rather not `LOCK(cs_main)` on every `BlockConnected` if avoidable. We wouldn't even need the lock in this method if `SetBestChain` was called with the tip block index as an argument. 3) I think this gives better efficiency of block locator hits. So if the tip of the block locator is always the best block index entry that has been flushed to disk, when the TxIndex restarts, it always picks up from the best possible point. Whereas if the locator is written on every BlockConnected call, because there are exponentially increasing jumps in the locator, you might only find a block that is further back than that one. (Hopefully that makes sense).

jamesob reviewed (approved):
Tested ACK https://github.com/bitcoin/bitcoin/pull/11857/commits/ea8be45ace75b649584c163deca3051c4f33aa16

This change is really nicely implemented. My comments are non-blocking - the code as-is reads well and represents a big improvement. `txindex` usage while testing this branch has been pleasant relative to its normally fickle behavior; I like that the index doesn't get blown away when toggling `txindex=`.

Here's the manual test plan I followed:

- [x] Start migration of existing (legacy) `blocks/index` database, interrupt it (on testnet)
  - [x] Continue interrupted migration of legacy `blocks/index` database (on testnet)
- [x] Build `indexes/txindex` from scratch (on testnet)
- [x] Retrieve a testnet transaction:
    ```sh
    $ which rpct

    rpct () {
      ./src/bitcoin-cli -rpcuser=foo -rpcpassword=xxx -rpcport=18332 "$@"
    }

    $ rpct getrawtransaction 9c86ecad5ecc447ef1540e13911f6b917b771d1780693613889f14efa7d6fc01

    020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff23030eb91304ac8bcf5a2f44504f4f4c2e544f502f484e2f0202f6d36693010000000000ffffffff028a910405000000001976a9140ac82b91aa7ff1b07c9673f28c8ced6c7045c41388ac0000000000000000266a24aa21a9eda00067c4917832fb72c8db625395da31e0e9e66d32d8385bf3d5bc1b036880e00120000000000000000000000000000000000000000000000000000000000000000000000000
    ```
- [x] After building `indexes/txindex`, retrieve a transaction from [a recent testnet block](https://live.blockcypher.com/btc-testnet/block/00000000000001cf81d8715947d225734d01fc2ff6f98257a745c516ac5b11d2/): 
    ```sh
     $ rpct getrawtransaction $(rpct getblock $(rpct getbestblockhash) | jq -r ".tx[0]")

     020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff230310b913042791cf5a2f44504f4f4c2e544f502f484e2f0202f6d3dea9010000000000ffffffff029d61af04000000001976a9140ac82b91aa7ff1b07c9673f28c8ced6c7045c41388ac0000000000000000266a24aa21a9edbdf213b3f829be29346dece981a7b90bad893358a0d7d95163887b6960760d150120000000000000000000000000000000000000000000000000000000000000000000000000
     ```
- [x] After building `indexes/txindex`, retrieve a transaction from [an older testnet block](https://live.blockcypher.com/btc-testnet/block/000000000000015db03f2c2690e2479a5a9a675298a76df0d759f04bd45bd71b/):
    ```sh
     $ rpct getrawtransaction $(rpct getblock $(rpct getblockhash 2) | jq -r ".tx[0]")
     
     01000000010000000000000000000000000000000000000000000000000000000000000000ffffffff0e0432e7494d010e062f503253482fffffffff0100f2052a010000002321038a7f6ef1c8ca0c588aa53fa860128077c9e6c11e6830f4d7ee4e763a56b7718fac00000000
     ```
- [x] Flip txindex from `1` to `0` and back to `1`, ensure full reindexing doesn't happen (on mainnet)
- [x] Do migration of existing legacy `blocks/index` database (on mainnet)
- [x] After building `indexes/txindex`, set txindex=0, restart, and run (on mainnet)
    ```sh
    $ rpc getrawtransaction 1242739fa76fe5f84fdcf2a7557cd059bc89969a0e9b43ded8f01a97ecd8f2db

    020000000001010000000000000000000000000000000000000000000000000000000000000000ffffffff4b03c7e607048413cf5a622f4254432e434f4d2ffabe6d6d37bccddf8230a0579980821ee6ad826521097fa2f3041c7492c1988eab282fc7010000000000000038014b294f56000000000000ffffffff02502ddd4a000000001976a91478ce48f88c94df3762da89dc8498205373a8ce6f88ac0000000000000000266a24aa21a9ed5a272ba104178bac1a18f9af13f1f13a8f727d2726f58f582d08d4b7076b8b7f0120000000000000000000000000000000000000000000000000000000000000000000000000

    $ rpc getrawtransaction 1242739fa76fe5f84fdcf2a7557cd059bc89969a0e9b43ded8f01a97ecd8f2db | sha256sum

    7ef9d4596788ca7bbc51bc0e82a0588c78d0115af91d3f73632085716af10cbf  -
    ```
- [x] Set txindex back to 1, restart, ensure no full reindexing happens, and run (on mainnet)
    ```sh
    $ rpc getrawtransaction 1242739fa76fe5f84fdcf2a7557cd059bc89969a0e9b43ded8f01a97ecd8f2db | sha256sum

    7ef9d4596788ca7bbc51bc0e82a0588c78d0115af91d3f73632085716af10cbf  -
    ``` 
- [x] Run `-reindex` after having built `indexes/txindex` (on testnet)

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +425,171 @@ bool CCoinsViewDB::Upgrade() {
+    // If the former, return value should still be true.
+    if (!Exists(DB_BEST_BLOCK)) {
+        locator.SetNull();
+        return true;
```
Because we successfully read from the database -- there was just no locator written, which in this case means no blocks have been written to the index yet.

jimpo commented:
@TheBlueMatt Can you please take a look at the last few PRs?

TheBlueMatt commented:
The fixup commits look fine, I think, except that you should probably add a commit to FlushStateToDisk to call SetBestChain in sync with fDoFullFlush always instead of it having its own timer. This shouldn't effect the wallet as in !fDoFullFlush cases we'll be replaying the blocks on restart anyway, but for this stuff would otherwise mean you still have a race.

Will do a full re-review when things are squashed/rebased.

jonasschnelli commented:
Needs rebase

TheBlueMatt reviewed (commented):
utACK 523dd763bfe39150e369332e415215d33f2a3ef5 despite the comments. I think its correct as-is now, but there's a few things that I think are rough API edges that would be nice to get cleanup on.

jimpo commented on src/txdb.cpp:
```diff
@@ -22,6 +22,7 @@ static const char DB_COIN = 'C';
 static const char DB_COINS = 'c';
 static const char DB_BLOCK_FILES = 'f';
 static const char DB_TXINDEX = 't';
+static const char DB_TXINDEX_BLOCK = 'T';
```
Suggest one.

jimpo commented on src/txdb.cpp:
```diff
@@ -424,3 +424,41 @@ bool CCoinsViewDB::Upgrade() {
+
+    // Read might have failed either because key does not exist or due to an error.
+    // If the former, return value should still be true.
+    if (!Exists(DB_BEST_BLOCK)) {
```
I'll change it, but I think the return value of `CDBWrapper::Read()` is super confusing on it's own. It means either that the key was not found or that the value at that key could not be deserialized into the provided struct.

jimpo commented on src/txdb.cpp:
```diff
@@ -462,3 +463,138 @@ bool TxIndexDB::WriteBestBlock(const CBlockLocator& locator)
+        batch_olddb.Erase(key);
+
+        if (batch_newdb.SizeEstimate() > batch_size || batch_olddb.SizeEstimate() > batch_size) {
+            WriteTxIndexMigrationBatches(*this, block_tree_db,
```
Seems like overkill to me to leave a comment for documented LevelDB behavior, but sure.

jimpo commented:
Closing and re-opening in #13033 because this has stopped loading for people.
